hecate_lexer/src/lib.rs

```diff
@@ -1 +1,225 @@
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(DivAssign);
+                } else {
+                    token_type = Token::Operator(Div);
+                }
+            }
+            &'>' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Ge);
+                } else {
+                    token_type = Token::Operator(Gt);
+                }
+            }
+            &'<' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Le);
+                } else {
+                    token_type = Token::Operator(Lt);
+                }
+            }
+            &'=' => {
+                if self.followed_by_equalsign() {
+                    token_type = Token::Operator(Eq);
+                } else {
+                    token_type = Token::Assignment;
+                }
+            }
+            &'&' => {
+                if self.iter.peek().unwrap() == &'&' {
+                    self.advance();
+                    token_type = Token::Operator(And);
+                } else {
+                    token_type = Token::Undefined;
+                    self.collect_erroneous();
+                }
+            }
+            &'|' => {
+                if self.iter.peek().unwrap() == &'|' {
+                    self.advance();
+                    token_type = Token::Operator(Or);
+                } else {
+                    token_type = Token::Undefined;
+                    self.collect_erroneous();
+                }
+            }
+            &'(' | &')' | &'{' | &'}' | &':' | &';' => {
+                self.advance();
+                token_type = Token::Delimiter;
+            }
+            _ => {
+                // Should be handled by parser
+                token_type = Token::Undefined;
+                self.collect_erroneous();
+            }
+        }
+
+        let span = Span::Span {
+            source: Source::String,
+            content: &self.input,
+            start: token_start,
+            end: self.current_pos,
+        };
+
+        let token = Spanned {
+            inner: token_type,
+            loc: span,
+        };
+
+        Option::Some(token)
+    }
+}
+
+impl<'a> Lexer<'a> {
+    fn collect_identifier(&mut self) {
+        self.advance();
+
+        while let Some(ch) = self.iter.peek() {
+            if ch.is_alphanumeric() || ch == &'_' {
+                self.advance();
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn collect_literal(&mut self) {
+        self.advance();
+
+        while let Some(ch) = self.iter.peek() {
+            if ch.is_alphanumeric() || ch == &'.' {
+                self.advance();
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn collect_erroneous(&mut self) {
+        self.advance();
+
+        while let Some(ch) = self.iter.peek() {
```

Comment:
what is a good default collection len for erroneous token? one char? a whole word? until whitespace? difficult to predict, not that vital as it ensures an error either way but we need a sensible default.

Reply:
maybe until either a whitespace is encountered or a known operator? I'm not sure, but is a important API definition.

Reply:
approaches:
a) greedily: until whitespace
  - easy
  - may catch too much in optional whitespace contextss
b) conservatively: just one char
  - also easy
  - may catch too little context for accurate error messages
c) until known
  - just join all unknowns together
  - fairly difficult when you want to avoid duplicate definitions of known tokens maybe?
  - would need some sort of intermediate aggregator?
c) contextual: until context switch (my personal fav)
  - would collect until border identifer -> control symbol (or) control symbol -> identifier
  - (or whitespace obviously)
  - should give most helpful diagnostic as we are neither over nor undergathering

Reply:
Could you elaborate on this? I am not sure what you mean with symbol -> identifier.